---
title: "Benchmark De methods"
author: "Enrico Gaffo"
date: "`r Sys.Date()`"
output: 
  html_document: 
    toc: yes
    toc_float: yes
    code_folding: hide
params:
  outdir: "benchmark_results/MS"
  inputData: "filtered_simdata/MS/datasetList.qs"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(SummarizedBenchmark)
library(magrittr)
# library(plotROC)
library(data.table)
library(qs)
library(DT)

# library(limma)
# library(edgeR)
# library(DESeq2)
# library(tximport)
# library(glmGamPoi)
# library(zinbwave)
library(scran)

library(BiocParallel)
library(batchtools)

# install.packages("devtools", Ncpus = multicoreWorkers())
# library(devtools)
# install_github("lichen-lab/circMeta", Ncpus = multicoreWorkers())
library(circMeta)
# library(lncDIFF)
# library(ShrinkBayes)
# library(samr)
set.seed(12345)
```

```{r}
source("../utils/diff_exp_funcs.R")
```

```{r}
## set the output directory
# outdir <- "benchmark_results"
outdir <- params$outdir
dir.create(path = outdir, recursive = T, showWarnings = F)
```

```{r}
nWorkers <- multicoreWorkers()

## load simulated preprocessed data
# inputData <- "./preprocessed_datasets/datasetList.qs"
inputData <- params$inputData
datasetList <- qread(file = inputData, nthreads = nWorkers)
datasetList <- datasetList[grepl("bulk", names(datasetList))]
```

```{r}
## separate DE and not-DE simulations
# datasetList <- datasetList[!grepl("mock", names(datasetList))]
## names(de_ds_list) <- paste0(names(de_ds_list), "_de")
# not_de_ds_list <- sim_ds_list[grepl("mock", names(sim_ds_list))]
```

# Prepare dataset and input parameters

```{r}
# subsetIdx <- c(1:3, 31:33, 61:63, # { bulk 3, 5, 10 } x 3
#                91:93, 121:123, 151:153) # { sice 3, 5, 10 } x 3
# subsetIdx <- c(1:2, 31:32, 61:62, 91:92, 121:122, 151:152)
# subsetIdx <- 1:2
# subsetIdx <- 1:length(datasetList) # all
subsetIdx <- which(grepl(pattern = "_bulk_", names(datasetList))) ## do not consider "sice" simulations
datasetList <- datasetList[subsetIdx]

# countData <- datasetList$N03_bulk_de_01$cntdat
# group <- datasetList$N03_bulk_de_01$coldat$condition
```

```{r}
## set scheduler for cluster computing
zwcpus <- 1 #8

walltime <-  86400 # 24h; 14400 4h max time
minmem <- 2048 # 4096 = 4Gb, 2048 = 2GByte, 8192 = 8Gb #Minimum memory required per allocated CPU.
bpparam <- BatchtoolsParam(workers = length(datasetList),
                           saveregistry = F,
                           cluster = "slurm",
                           resources = list(ncpus = zwcpus,
                                            walltime = walltime,
                                            memory = minmem)
)

## Multithread, single machine
# bpparam <- MulticoreParam(nWorkers)
# zwcpus <- 1

## set max runtime allowed (set a lower value than the scheduler)
# bptimeout(bpparam) <- walltime - 10
```

## Compute weigths

```{r prepare_datasets}
## prepare dataset and input parameters
datasetList <- 
    bplapply(datasetList, 
             function(x, compute_weights, zwcpus) {
                 
                 inner_bpparam <- BiocParallel::MulticoreParam(workers = zwcpus)
                 
                 ## keep track of runtime spent computing weights with ZinbWave; 
                 ## it will be sum up to the running time of methods using weights
                 tictoc::tic(msg = "Computing weights")
                 weights <- compute_weights(countdata = x$cntdat,
                                            coldat = x$coldat[[1]],
                                            BPPARAM = inner_bpparam)
                 # weights = NULL, ## only for quick testing. Uncomment above for real weights
                 time_weights <- tictoc::toc(log = F, quiet = T)
                 
                 x$weights <- weights
                 x$time_weights <- setNames(rep(time_weights$toc - time_weights$tic, nrow(x$cntdat)), NULL)
                 x
                 
             }, 
             
             compute_weights = compute_weights,
             zwcpus = zwcpus,
             BPPARAM = bpparam)
```

```{r save_datasets}
# ## save the processed data sets
# inputData <- "datasetList.qs"
# qsave(x = datasetList, 
#       file = inputData, 
#       nthreads = multicoreWorkers(), 
#       preset = "fast")
```

# Bench design

```{r}
bd <- BenchDesign()
```

## Add memthods to the bench

```{r}
## add memthods to the bench
bd <- bd %>%
    addMethod(label = "DESeq2_Dt_WaT", ## DESeq2, default parameters, Wald test
              func = deseq2_run, 
              # post = deseq2_pv,
              post = list(pv = deseq2_pv,
                          adj_pv = deseq2_apv,
                          lfc = deseq2_lfc,
                          runtime = deseq2_time),
              meta = list(pkg_name = "DESeq2", pkg_vers = as.character(packageVersion("DESeq2"))),
              params = rlang::quos(countData = cntdat,
                                   colData = coldat, 
                                   design = ~condition,
                                   contrast = c("condition", "2", "1"))) %>%
    addMethod(label = "DESeq2_Dt_LRT", ## DESeq2, default parameters, likelihood ratio test
              func = deseq2lrt_run, 
              # post = deseq2_pv,
              post = list(pv = deseq2_pv,
                          adj_pv = deseq2_apv,
                          lfc = deseq2_lfc,
                          runtime = deseq2_time),
              meta = list(pkg_name = "DESeq2", pkg_vers = as.character(packageVersion("DESeq2"))),
              params = rlang::quos(countData = cntdat,
                                   colData = coldat, 
                                   design = ~condition,
                                   contrast = c("condition", "2", "1"))) %>%
    addMethod(label = "DESeq2_Lc_LRT", ## DESeq2, parameters for low counts, likelihood ratio test
              func = deseq2lc_run, 
              # post = deseq2_pv,
              post = list(pv = deseq2_pv,
                          adj_pv = deseq2_apv,
                          lfc = deseq2_lfc,
                          runtime = deseq2_time),
              meta = list(pkg_name = "DESeq2", pkg_vers = as.character(packageVersion("DESeq2"))),
              params = rlang::quos(countData = cntdat,
                                   colData = coldat, 
                                   design = ~condition,
                                   contrast = c("condition", "2", "1"))) %>%
    addMethod(label = "DESeq2_Zi_LRT", ## DESeq2, parameters for single-cell data, likelihood ratio test
              func = deseq2zi_run, 
              # post = deseq2_pv, 
              post = list(pv = deseq2_pv,
                          adj_pv = deseq2_apv,
                          lfc = deseq2_lfc,
                          runtime = deseq2_time),
              meta = list(pkg_name = "DESeq2", pkg_vers = as.character(packageVersion("DESeq2"))),
              params = rlang::quos(countData = cntdat,
                                   colData = coldat, 
                                   design = ~condition,
                                   contrast = c("condition", "2", "1"))) %>%
    addMethod(label = "DESeq2_Sc_LRT", ## DESeq2, parameters for single-cell data and scran::computeSumFactors, likelihood ratio test
              func = deseq2sc_run, 
              # post = deseq2_pv,
              post = list(pv = deseq2_pv,
                          adj_pv = deseq2_apv,
                          lfc = deseq2_lfc,
                          runtime = deseq2_time),
              meta = list(pkg_name = "DESeq2/scran",
                          pkg_vers = paste(as.character(packageVersion("DESeq2")),
                                           as.character(packageVersion("scran")),
                                           sep = "/")),
              params = rlang::quos(countData = cntdat,
                                   colData = coldat,
                                   design = ~condition,
                                   contrast = c("condition", "2", "1"))) %>%
    addMethod(label = "DESeq2_GP_LRT", ## DESeq2, Gamma-Poisson, likelihood ratio test
              func = deseq2gpLrt_run, 
              # post = deseq2_pv, 
              post = list(pv = deseq2_pv,
                          adj_pv = deseq2_apv,
                          lfc = deseq2_lfc,
                          runtime = deseq2_time),
              meta = list(pkg_name = "DESeq2/glmGamPoi", 
                          pkg_vers = paste(as.character(packageVersion("DESeq2")), 
                                           as.character(packageVersion("glmGamPoi")), 
                                           sep = "/")),
              params = rlang::quos(countData = cntdat,
                                   colData = coldat, 
                                   design = ~condition,
                                   contrast = c("condition", "2", "1"))) %>%
    addMethod(label = "DESeq2_ZW_LRT", ## DESeq2, ZinbWave, likelihood ratio test
              func = deseq2zw_run, 
              # post = deseq2_pv, 
              post = list(pv = deseq2_pv,
                          adj_pv = deseq2_apv,
                          lfc = deseq2_lfc,
                          runtime = deseq2_time),
              meta = list(pkg_name = "DESeq2/zinbwave", 
                          pkg_vers = paste(as.character(packageVersion("DESeq2")), 
                                           as.character(packageVersion("zinbwave")), 
                                           sep = "/")),
              params = rlang::quos(countData = cntdat,
                                   colData = coldat, 
                                   design = ~condition,
                                   contrast = c("condition", "2", "1"),
                                   weights = weights)) %>%
    addMethod(label = "DESeq2_BP_WaT", ## DESeq2, betaPrior=T, Wald test
              func = deseq2bp_run, 
              # post = deseq2_pv,
              post = list(pv = deseq2_pv,
                          adj_pv = deseq2_apv,
                          lfc = deseq2_lfc,
                          runtime = deseq2_time),
              meta = list(pkg_name = "DESeq2", pkg_vers = as.character(packageVersion("DESeq2"))),
              params = rlang::quos(countData = cntdat,
                                   colData = coldat, 
                                   design = ~condition,
                                   contrast = c("condition", "2", "1"))) %>%
    addMethod(label = "edgeR_Dt_LRT", ## edgeR, defaults, likelihood ratio test 
              func = edgeR_run, 
              # post = edgeR_pv,
              post = list(pv = edgeR_pv,
                          adj_pv = edgeR_apv,
                          lfc = edgeR_lfc,
                          runtime = edgeR_time),
              meta = list(pkg_name = "edgeR", pkg_vers = as.character(packageVersion("edgeR"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition)) %>%
    addMethod(label = "edgeR_ZW_LRT", ## edgeR, ZinbWave, likelihood ratio test 
              func = edgeRzw_run, 
              # post = edgeR_pv,
              post = list(pv = edgeR_pv,
                          adj_pv = edgeR_apv,
                          lfc = edgeR_lfc,
                          runtime = edgeR_time),
              meta = list(pkg_name = "edgeR/zinbwave", 
                          pkg_vers = paste(as.character(packageVersion("edgeR")), 
                                           as.character(packageVersion("zinbwave")), 
                                           sep = "/")),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition,
                                   weights = weights)) %>%
    addMethod(label = "edgeR_rbst_LRT", ## edgeR, robust dispersion estimate, likelihood ratio test 
              func = edgeRrbst_run, 
              # post = edgeR_pv,
              post = list(pv = edgeR_pv,
                          adj_pv = edgeR_apv,
                          lfc = edgeR_lfc,
                          runtime = edgeR_time),
              meta = list(pkg_name = "edgeR", pkg_vers = as.character(packageVersion("edgeR"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition)) %>%
    addMethod(label = "edgeR_rbst50df_LRT", ## edgeR, robust dispersion estimate with 50 degrees of freedom, likelihood ratio test 
              func = edgeRrbst50df_run, 
              # post = edgeR_pv,
              post = list(pv = edgeR_pv,
                          adj_pv = edgeR_apv,
                          lfc = edgeR_lfc,
                          runtime = edgeR_time),
              meta = list(pkg_name = "edgeR", pkg_vers = as.character(packageVersion("edgeR"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition)) %>%
    addMethod(label = "edgeR_rbstEdf_LRT", ## edgeR, robust dispersion estimate with autodetermied number of degrees of freedom, likelihood ratio test 
              func = edgeRrbstEdf_run, 
              # post = edgeR_pv,
              post = list(pv = edgeR_pv,
                          adj_pv = edgeR_apv,
                          lfc = edgeR_lfc,
                          runtime = edgeR_time),
              meta = list(pkg_name = "edgeR", pkg_vers = as.character(packageVersion("edgeR"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition)) %>%
    addMethod(label = "edgeR_rbst_QFT", ## edgeR, robust dispersion estimate, empirical Bayes quasi-likelihood F-test 
              func = edgeRql_run, 
              post = list(pv = edgeR_pv,
                          adj_pv = edgeR_apv,
                          lfc = edgeR_lfc,
                          runtime = edgeR_time),
              meta = list(pkg_name = "edgeR", pkg_vers = as.character(packageVersion("edgeR"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition)) %>%
    addMethod(label = "edgeR_rbstTMMswp_LRT",
              func = edgeRrbstTMMswp_run,
              post = list(pv = edgeR_pv,
                          adj_pv = edgeR_apv,
                          lfc = edgeR_lfc,
                          runtime = edgeR_time),
              meta = list(pkg_name = "edgeR", 
                          pkg_vers = as.character(packageVersion("edgeR"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition)) %>%
    addMethod(label = "voom_Dt_MFT", ## limma-voom, defauls, moderated F-statistic (or is it t-stats ?) 
              func = voom_run, 
              # post = voom_pv,
              post = list(pv = voom_pv,
                          adj_pv = voom_apv,
                          lfc = voom_lfc,
                          runtime = voom_time),
              meta = list(pkg_name = "limma-voom", pkg_vers = as.character(packageVersion("limma"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition)) %>%
    addMethod(label = "voom_Rb_MFT", ## limma-voom, use robust est, moderated F-statistic (or is it t-stats ?) 
              func = voomRbst_run, 
              # post = voom_pv,
              post = list(pv = voom_pv,
                          adj_pv = voom_apv,
                          lfc = voom_lfc,
                          runtime = voom_time),
              meta = list(pkg_name = "limma-voom", pkg_vers = as.character(packageVersion("limma"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition)) %>%
    addMethod(label = "voom_Qn_MFT", ## limma-voom, use quantile normalization, moderated F-statistic (or is it t-stats ?) 
              func = voomQn_run, 
              # post = voom_pv,
              post = list(pv = voom_pv,
                          adj_pv = voom_apv,
                          lfc = voom_lfc,
                          runtime = voom_time),
              meta = list(pkg_name = "limma-voom", pkg_vers = as.character(packageVersion("limma"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition)) %>%
    addMethod(label = "voom_Sp_MFT", ## limma-voom, defaults as in vignette, moderated F-statistic (or is it t-stats ?) 
              func = voomSimple_run, 
              # post = voom_pv,
              post = list(pv = voom_pv,
                          adj_pv = voom_apv,
                          lfc = voom_lfc,
                          runtime = voom_time),
              meta = list(pkg_name = "limma-voom", pkg_vers = as.character(packageVersion("limma"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition)) %>%
    addMethod(label = "voom_LF_MFT", ## limma-voom, use voomLmFit, moderated F-statistic (or is it t-stats ?) 
              func = voomLmFit_run, 
              # post = voom_pv,
              post = list(pv = voom_pv,
                          adj_pv = voom_apv,
                          lfc = voom_lfc,
                          runtime = voom_time),
              meta = list(pkg_name = "limma-voom", pkg_vers = as.character(packageVersion("limma"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition)) %>%
    addMethod(label = "voom_ZW_MFT", ## limma-voom, use ZinbWave weigths, moderated F-statistic (or is it t-stats ?) 
              func = voomZw_run, 
              # post = voom_pv,
              post = list(pv = voom_pv,
                          adj_pv = voom_apv,
                          lfc = voom_lfc,
                          runtime = voom_time),
              meta = list(pkg_name = "limma-voom/zinbwave", 
                          pkg_vers = paste(as.character(packageVersion("limma")), 
                                           as.character(packageVersion("zinbwave")), 
                                           sep = "/")),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition,
                                   design = ~coldat$condition,
                                   weights = weights)) %>%
    addMethod(label = "circMeta_Dt_PZT", ## circMeta, defaults, Poisson z-test
              func = cMeta_run, 
              post = list(pv = cMeta_pv,
                          adj_pv = cMeta_apv,
                          lfc = cMeta_lfc,
                          runtime = cMeta_time),
              meta = list(pkg_name = "circMeta", pkg_vers = as.character(packageVersion("circMeta"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "circMeta_Lc_PZT", ## circMeta, sqrt of counts, Poisson z-test
              func = cMetaLc_run, 
              post = list(pv = cMeta_pv,
                          adj_pv = cMeta_apv,
                          lfc = cMeta_lfc,
                          runtime = cMeta_time),
              meta = list(pkg_name = "circMeta", pkg_vers = as.character(packageVersion("circMeta"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "lncDIFF_Dt_LRT", 
              func = lncdiff_run, 
              post = list(pv = lncdiff_pv,
                          adj_pv = lncdiff_apv,
                          lfc = lncdiff_lfc,
                          runtime = lncdiff_time),
              meta = list(pkg_name = "lncDIFF", 
                          pkg_vers = as.character(packageVersion("lncDIFF"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "SAMseq_Dt_TT", 
              func = samseq_run,
              post = list(pv = samseq_pv,
                          adj_pv = samseq_apv,
                          lfc = samseq_lfc,
                          runtime = samseq_time),
              meta = list(pkg_name = "samr", 
                          pkg_vers = as.character(packageVersion("samr"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "NBID_Dt_LRT", 
              func = nbid_run,
              post = list(pv = nbid_pv,
                          adj_pv = nbid_apv,
                          lfc = nbid_lfc,
                          runtime = nbid_time),
              meta = list(pkg_name = "NBID", 
                          pkg_vers = as.character(packageVersion("NBID"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "NBID_Sc_LRT", 
              func = nbid_sc_run,
              post = list(pv = nbid_pv,
                          adj_pv = nbid_apv,
                          lfc = nbid_lfc,
                          runtime = nbid_time),
              meta = list(pkg_name = "NBID/scran", 
                          pkg_vers = paste(as.character(packageVersion("NBID")),
                                           as.character(packageVersion("scran")),
                                           sep = "/")),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "NOISeqBIO_TMM_LRT",
              func = noiseqbio_run,
              post = list(pv = noiseqbio_pv,
                          adj_pv = noiseqbio_apv,
                          lfc = noiseqbio_lfc,
                          runtime = noiseqbio_time),
              meta = list(pkg_name = "NOISeqBIO",
                          pkg_vers = as.character(packageVersion("NOISeq"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat)) %>%
    addMethod(label = "PoissonSeq_Dt_TT",
              func = poissonSeq_run,
              post = list(pv = poissonSeq_pv,
                          adj_pv = poissonSeq_apv,
                          lfc = poissonSeq_lfc,
                          runtime = poissonSeq_time),
              meta = list(pkg_name = "PoissonSeq",
                          pkg_vers = as.character(packageVersion("PoissonSeq"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "limma_St_Vst",
              func = limmaStVst_run,
              post = list(pv = limmaStVst_pv,
                          adj_pv = limmaStVst_apv,
                          lfc = limmaStVst_lfc,
                          runtime = limmaStVst_time),
              meta = list(pkg_name = "LimmaVst",
                          pkg_vers = paste(as.character(packageVersion("limma")),
                                           as.character(packageVersion("sctransform")),
                                           sep = "/")),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "ROTScpm_Dt_TLT", ## t-like test statistics
              func = ROTScpm_run,
              post = list(pv = ROTScpm_pv,
                          adj_pv = ROTScpm_apv,
                          lfc = ROTScpm_lfc,
                          runtime = ROTScpm_time),
              meta = list(pkg_name = "ROTScpm",
                          pkg_vers = paste(as.character(packageVersion("ROTS")),
                                           as.character(packageVersion("edgeR")),
                                           sep = "/")),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "DEsingle_Dt_LRT",
              func = DEsingle_run,
              post = list(pv = DEsingle_pv,
                          adj_pv = DEsingle_apv,
                          lfc = DEsingle_lfc,
                          runtime = DEsingle_time),
              meta = list(pkg_name = "DEsingle",
                          pkg_vers = as.character(packageVersion("DEsingle"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "metagenomeSeq_Dt_EBT",
              func = metagenomeSeq_run,
              post = list(pv = metagenomeSeq_pv,
                          adj_pv = metagenomeSeq_apv,
                          lfc = metagenomeSeq_lfc,
                          runtime = metagenomeSeq_time),
              meta = list(pkg_name = "metagenomeSeq",
                          pkg_vers = as.character(packageVersion("metagenomeSeq"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "seurat_Bim_LRT",
              func = seuratBimod_run,
              post = list(pv = seurat_pv,
                          adj_pv = seurat_apv,
                          lfc = seurat_lfc,
                          runtime = seurat_time),
              meta = list(pkg_name = "Seurat",
                          pkg_vers = as.character(packageVersion("Seurat"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "seurat_Dt_WLX",
              func = seuratWilcox_run,
              post = list(pv = seurat_pv,
                          adj_pv = seurat_apv,
                          lfc = seurat_lfc,
                          runtime = seurat_time),
              meta = list(pkg_name = "Seurat",
                          pkg_vers = as.character(packageVersion("Seurat"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "MAST_CPM_LRT",
              func = MASTcpm_run,
              post = list(pv = MASTcpm_pv,
                          adj_pv = MASTcpm_apv,
                          lfc = MASTcpm_lfc,
                          runtime = MASTcpm_time),
              meta = list(pkg_name = "MAST",
                          pkg_vers = as.character(packageVersion("MAST"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "monocle_Dt_LRT",
              func = monocle_run,
              post = list(pv = monocle_pv,
                          adj_pv = monocle_apv,
                          lfc = monocle_lfc,
                          runtime = monocle_time),
              meta = list(pkg_name = "monocle",
                          pkg_vers = as.character(packageVersion("monocle"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition)) %>%
    addMethod(label = "wilcoxon_TMM_WLX",
              func = wilcoxon_run,
              post = list(pv = wilcoxon_pv,
                          adj_pv = wilcoxon_apv,
                          lfc = wilcoxon_lfc,
                          runtime = wilcoxon_time),
              meta = list(pkg_name = "stats",
                          pkg_vers = as.character(packageVersion("stats"))),
              params = rlang::quos(countData = cntdat,
                                   group = coldat$condition))

bd
```

```{r}
# printMethods(bd)
```

## Build bench list

```{r build_benches, message=FALSE}
## override BPPARAM
parallel_methods <- 1
# bpparam$resources$ncpus <- parallel_methods
walltime <- 3600 ## let max 1h computing time to each method

# BiocParallel::bptimeout(bpparam) <- 180 # 3min

sbL <- 
  bplapply(datasetList, 
           function(x, bd, parallel_methods, walltime = 1800) { 
             
             inner_bpparam <- BiocParallel::MulticoreParam(parallel_methods)
             BiocParallel::bptimeout(inner_bpparam) <- walltime
             # inner_bpparam <- BiocParallel::SerialParam()
             
             SummarizedBenchmark::buildBench(bd, data = x, 
                                             truthCols = c(pv = "status",
                                                           adj_pv = "status", 
                                                           lfc = "status", # LFC(DEC) >= 0.5
                                                           runtime = "time_weights"), 
                                             keepData = T,
                                             parallel = T, #F do not run in parallel 
                                             BPPARAM = inner_bpparam) # this is unused if parallel =F
           }, 
           bd = bd,
           parallel_methods = parallel_methods,
           walltime = walltime,
           BPPARAM = bpparam)
# sbL
```

```{r}
## error handling
show_dt <- 
  dcast(melt(rbindlist(lapply(sbL, 
                              function(x)data.table(as.data.frame(simplify2array(metadata(x)$sessions[[1]]$results)), 
                                                    keep.rownames = "Assay")), 
                       idcol = "DS"), 
             id.vars = c("DS", "Assay"), 
             variable.name = "Method"), 
        formula = DS + Method ~ Assay)

show_dt$DS <- factor(show_dt$DS)
show_dt$Method <- factor(show_dt$Method)

datatable(show_dt[adj_pv != "success" | 
                    lfc != "success" | 
                    pv != "success" | 
                    runtime != "success"], 
          caption = "Methods that failed",
          filter = "top", rownames = F)
```

```{r}
## save the summarizedBenchmark builds
sumBenchs_qs <- file.path(outdir, "sumBenchs.qs")
qsave(x = sbL, 
      file = sumBenchs_qs, 
      nthreads = multicoreWorkers(), 
      preset = "fast")
```

The built SummarizedBenchmark objects have been save into <a href="`r sumBenchs_qs`">`r sumBenchs_qs`</a>.  

# Performance

```{r}
# availableMetrics()
```

## Add performance 

```{r}
#' Adds performance metrics to a SummarizedBenchmark object
#' param x the SummarizedBenchmark object
#' returns a SummarizedBenchmark with the peformance metrics setted
add_performance_metrics <- 
  function(x) {
    
    ## Notes:
    ## precision (PPV) = 1 - FDR
    ## recall = TPR
    ## specificity = TNR  
    ## FPR = 1 - TNR
    ## FNR = 1 - TPR
    ## F1 = 2 * ( (PPV * TPR) / (PPV + TPR) )
    
    ## add the metrics on the P-values
    x <- SummarizedBenchmark::addPerformanceMetric(x, 
                                                   evalMetric = c("rejections", "TPR", "TNR", "FDR"), #, "FNR"
                                                   assay = "pv")
    
    ## add the metrics on the adjusted P-values
    x <- SummarizedBenchmark::addPerformanceMetric(x, 
                                                   evalMetric = c("rejections", "TPR", "TNR", "FDR"), #, "FNR"
                                                   assay = "adj_pv")
    
    ## add the metrics on the fold changes: TPR
    x <- SummarizedBenchmark::addPerformanceMetric(object = x,
                                                   assay = "lfc",
                                                   evalMetric = "LFC_TPR",
                                                   evalFunction = function(query, truth, lfc_thr = 0.5) {
                                                     ## TPR = TP / (TP + FN)
                                                     is_lfc_larger  <- query >= lfc_thr
                                                     is_lfc_larger[is.na(is_lfc_larger)] <- F
                                                     TP <- sum(is_lfc_larger & truth == 1)
                                                     TP / sum(truth == 1)
                                                   })
    
    ## add the metrics on the fold changes: TNR
    # x <- SummarizedBenchmark::addPerformanceMetric(object = x,
    #                                                assay = "lfc",
    #                                                evalMetric = "LFC_TNR", 
    #                                                evalFunction = function(query, truth, lfc_thr = 0.5) {
    #                                                  ## TNR = TN / N, with N = TN + FP
    #                                                  is_lfc_lower  <- query < lfc_thr
    #                                                  is_lfc_lower[is.na(is_lfc_lower)] <- T
    #                                                  TN <- sum(is_lfc_lower & truth == 0)
    #                                                  TN / sum(truth == 0)
    #                                                })
    # 
    # ## add the metrics on the fold changes: FDR
    # x <- SummarizedBenchmark::addPerformanceMetric(object = x,
    #                                                assay = "lfc",
    #                                                evalMetric = "LFC_FDR", 
    #                                                evalFunction = function(query, truth, lfc_thr = 0.5) {
    #                                                  ## FDR = FP / (FP + TP)
    #                                                  is_lfc_larger  <- query >= lfc_thr
    #                                                  is_lfc_larger[is.na(is_lfc_larger)] <- F
    #                                                  FP <- sum(is_lfc_larger & truth == 0)
    #                                                  TP <- sum(is_lfc_larger & truth == 1)
    #                                                  FP / (FP + TP)
    #                                                })
    
    ## add the Runtime metric
    x <- SummarizedBenchmark::addPerformanceMetric(object = x,
                                                   assay = "runtime",
                                                   evalMetric = "Runtime",
                                                   evalFunction = function(query, truth, add_weight_time = FALSE) {
                                                     ifelse(add_weight_time, query[1] + truth[1], query[1])
                                                   })
    
    ## return the updated SummarizedBenchmark object
    x
  }
```

```{r add_performance_metrics}
## add the performance metrics to the list of bechDesign
# bpparam$resources$ncpus <- 1

sbL <- 
  bplapply(sbL, 
           add_performance_metrics,
           BPPARAM = bpparam)
# BPPARAM = BiocParallel::MulticoreParam(nWorkers))
```

## Estimate performance

```{r estimate_performance}
alpha_targets <- c(0.01, 0.05, 0.1) 
add_weights <- c(FALSE, TRUE)

sbL <- bplapply(sbL, function(x, alphas, add_weights) {
    SummarizedBenchmark::estimatePerformanceMetrics(x, 
                                                    rerun = T,
                                                    alpha = alphas, 
                                                    add_weight_time = add_weights,
                                                    addColData = T)},
    alphas = alpha_targets,
    add_weights = add_weights,
    # BPPARAM = BiocParallel::MulticoreParam(nWorkers))
    BPPARAM = bpparam)

# View(rbindlist(lapply(sbL,
#                       function(x)data.table(as.data.frame(colData(x)),
#                                             keep.rownames = "Met")),
#                idcol = "DS"))
```

```{r save_performance_results}
## save the summarizedBenchmark estimated performance
sumBench_perf_metrics_qs <- file.path(outdir, "sumBench_perf_metrics.qs")
qsave(x = sbL, 
      file = sumBench_perf_metrics_qs, 
      nthreads = multicoreWorkers(), 
      preset = "fast")
```

The benchmark results have been save into <a href="`r sumBench_perf_metrics_qs`">`r sumBench_perf_metrics_qs`</a>.  

# Session info

```{r}
sessionInfo()
```

